chore(docs): document Igor's behavior and features in README.md (#492)
spinnaker/igor: README.md

> Igor is a service that provides a single point of integration with CI and SCM services for Spinnaker.

Reviewer: ```suggestion
Igor is a service that provides a single point of integration with Continuous Integration (CI) and Source Control Management (SCM) services for Spinnaker.
```

# Common Polling Architecture

Igor runs a number of pollers that all share the same common architecture. At a high level, they all:

- periodically get a list of items from an external resource (e.g. builds on a Jenkins master)
- compare that list against their own persisted cache of items (the difference is called _delta size_)
- send an `echo` event for each new item

> - caches the new list of items

Reviewer: ```suggestion
- cache the new list of items
```

Features:

- *locking*: pollers can optionally acquire a distributed lock in the storage system before attempting to complete a polling cycle. This makes it possible to run Igor in a high-availability configuration and scale it horizontally.

> - *safe guards*: abnormally large delta sizes can indicate a problem (e.g. lost or corrupt cache data) and cause downstream issues. If a polling cycle results in a delta size above the threshold

Reviewer 2: i am sure you wanted to say something else after `above the threshold`

Reviewer 2: Also, maybe worth mentioning the `admin` EP to bypass this safeguard?
`admin/pollers/fastforward/{monitorName}`

Author: Haha good catch! I was probably actually wondering what does happen we're above the threshold and then got sidetracked, refrigerator

> - *safe guards*: abnormally large delta sizes can indicate a problem (e.g. lost or corrupt cache data) and cause downstream issues. If a polling cycle results in a delta size above the threshold

Reviewer: ```suggestion
- *safeguards*: abnormally large delta sizes can indicate a problem (e.g. lost or corrupt cache data) and cause downstream issues. If a polling cycle results in a delta size above the threshold
```

- *health*: Igor has a `HealthIndicator` that reports `Down` if no pollers are running or if they have not had a successful polling cycle in a long time

Relevant properties:

| *Property* | *Default value* | *Description* |
| --- | --- | --- |
| `spinnaker.build.pollInterval` | `60` | Interval in seconds between polling cycles |
| `spinnaker.pollingSafeguard.itemUpperThreshold` | `1000` | Defines the upper threshold for number of new items before a cache update cycle will be rejected | `locking.enabled` | `false` | Enables distributed locking so that Igor can run on multiple nodes without interference |

Relevant metrics:

> | *Property* | *Type* | *Description* |

Reviewer 2: probably
```suggestion
| *Metric* | *Type* | *Description* |
```

| --- | --- | --- |
| `pollingMonitor.newItems` | gauge | represents the number of new items cached by a given monitor during a polling cycle |
| `pollingMonitor.itemsOverThreshold` | gauge | 0 if deltaSize < threshold, deltaSize otherwise |

> | `pollingMonitor.pollTiming` | timer | published for every polling cycle with the duration it took to complete |

Reviewer 2: do you think it's worth mentioning this event is fired with the name of the monitor as a tag `monitor`

# Storage

The following storage backends are supported:

- Redis
- Dynomite

Relevant properties:
```
redis:
  enabled: true
  connection: redis://host:port
```

# Integration with SCM services

The following SCM services are supported:

- Stash
- Github
- Gitlab
- Bitbucket

`Commit` controller classes expose APIs to retrieve lists of commits, such as `/github/{{projectKey}}/{{repositorySlug}}/compareCommits?from={{fromHash}}&to={{toHash}}`

At the moment Igor only exposes read APIs, there are no pollers and no triggers involving SCM services directly.

Relevant properties:

```
github:
  baseUrl: "https://api.github.com"
  accessToken: '<your github token>'
  commitDisplayLength: 8

stash:
  baseUrl: "<stash url>"
  username: '<stash username>'
  password: '<stash password>'

bitbucket:
  baseUrl: "https://api.bitbucket.org"
  username: '<bitbucket username>'
  password: '<bitbucket password>'
  commitDisplayLength: 7

gitlab:
  baseUrl: "https://gitlab.com"
  privateToken: '<your gitlab token>'
  commitDisplayLength: 8
```

# Integration with CI services

The following CI services are supported:

- Jenkins
- Gitlab CI
- Wercker
- Travis
- Artifactory

> - Concourse

Reviewer: Google Cloud Build (GCB) is also supported.  (Though it is a bit different in that it doesn't poll GCB and requires setting up pubsub subscriptions as per [here](https://www.spinnaker.io/setup/ci/gcb/)).  Also maybe it makes sense to alphabetize the list?

Author: Ah, I didn't know about GCB. I was literally looking for subclasses of `CommonPollingMonitor` :)

Good point about the ordering 👍
